### ISSUE TITLE
Request for Comment: Clarify Our Curricular Guidelines

### ISSUE THREAD

**Issue by waciumawanjohi** [ossu/computer-science#679]
**Problem**:
OSSU references a wide variety of curricula and guidelines, making course evaluation more difficult.

**Duration**:
2020/04/20

**Background**:
OSSU’s tagline is simple and clear. To provide a `Path to a free self-taught education in Computer Science!` Our summary further clarifies that `It is designed according to the degree requirements of undergraduate computer science majors, minus general education (non-CS) requirement[s]`.

At the same time, OSSU has three different sections that discuss our [academic requirements](https://github.com/ossu/computer-science#references), [references](https://github.com/ossu/computer-science#references) or [curricular guidelines](https://github.com/ossu/computer-science/blob/master/extras/curriculum_guidelines.md). Even within these, there are multiple references. Curriculum Guidelines refers to both the ABET program criteria and the ACM/IEEE curriculum guidelines. If these conflict, which should be paramount?

Only one of the resources in all of these sections aligns with OSSU’s mission: CS2013, the Curriculum Guidelines for Undergraduate Programs in Computer Science from ACM/IEEE. The goal of CS2013 is “to establish international curricular guidelines for undergraduate programs in computing.” The document establishes a number of principles to which its advice must adhere. These include:

* The CS2013 guidelines must be relevant to a variety of institutions.
* CS2013 should provide guidance for the expected level of mastery of topics by graduates.
* The size of the essential knowledge must be managed.

All are applicable to OSSU.

CS2013 provides clear expectations for a CS curriculum; which topics must be covered and to what extent. The level of detail provides a rubric that can be used to judge necessary changes to OSSU.

Multiple RFCs have already incorporated CS2013 advice as justification. In Issues it is the most referenced resource from the academic requirements, references and curricular guidelines. Codifying its primacy will make it clearer to new contributors how to evaluate resources for addition to OSSU.

**Proposal**:
* List CS2013 as the solo curricular guideline.
* Create a page for Alternate Curricula. 
    * List other self taught curriculum guides here (e.g. Teach Yourself CS, p1xt)
    * List ABET under a section for alternate curricular guidelines. Guidelines from other accreditation bodies (i.e. other countries) can similarly be included.
    * Make clear that these are high quality resources, but not our guideline documents.

**Alternatives**:
1) Move all resources to curriculum guidelines. This would not clarify our guidelines, but at least consolidates information in the same place.

2) Choose a different curricular guideline, perhaps ABET. There are a few challenges to this. First, ABET is underspecified. [CS2013 allows OSSU to judge whether prolog should be a required or elective course](https://github.com/ossu/computer-science/issues/669#issuecomment-599303642). ABET states programs must have, “Substantial coverage of algorithms and complexity, computer science theory, concepts of programming languages, and software development.” but does not go further in detail. Second, ABET focuses primarily on American universities, while CS2013 explicitly international in scope.

**Comment by amorriscode** [comment]
I definitely believe keeping things as simple as possible is a great approach.

**Comment by Alaharon123** [comment]
It occurs to me that despite talking to you outside of Github, I never actually made a comment on this issue. I should do that. I'll start by quoting the curricular guidelines of ABET so I can easily reference them.

> **Curriculum**
> The curriculum requirements specify topics, but do not prescribe specific courses.
> These requirements are:
> 
> a. Computer science: At least 40 semester credit hours (or equivalent) that must include:
> 1.  Substantial coverage of algorithms and complexity, computer science theory, concepts of programming languages, and software development.
> 2.  Substantial coverage of at least one general-purpose programming language.
> 3.  Exposure to computer architecture and organization, information management, networking and communication, operating systems, and parallel and distributed computing.
> 4.  The study of computing-based systems at varying levels of abstraction.
> 5.  A major project that requires integration and application of knowledge and skills acquired in earlier course work.­
> 
> b. Mathematics: At least 15 semester credit hours (or equivalent) that must include discrete mathematics and must have mathematical rigor at least equivalent to introductory calculus. The additional mathematics might include course work in areas such as calculus, linear algebra, numerical methods, probability, statistics, or number theory.
> c. [Science:] At least six semester credit hours (or equivalent) in natural science course work intended for science and engineering majors. This course work must develop an understanding of the scientific method and must include laboratory work.

ACM covers a1, a3, and a4. a2 is incredibly vague and I don't think we should worry about that because the most important thing is having high quality courses that cover everything. I think when we finish ensuring that the entirety of ACM is covered, we'll find that that has occurred naturally or we can say that that has occurred because who knows what the word substantial means anyway. Maybe two courses is enough even. a5 is currently covered with a final project. Should that project have more detailed requirements or anything like that? I don't know, but it *is* currently covered in some fashion. 

So what that leaves us with is the Math and Science part. Whether or not we should include Math and Science, I don't know, but I do think that it pays to make a proper decision on that. I'm certainly in favor of Single Variable Calculus being mandatory, but I have yet to find a good online Physics course and we'd be missing the lab work that ABET requires anyway.

**tl;dr**, I think we should ignore the CS part of the ABET guidelines for CS at least until we're satisfied with ACM coverage, and I think the primary decision should be on whether or not to include Math and Science. I'm not strongly in favor either way, but I think it's more important to make a decision than what that decision is.

As for a page for alternate curricula, sure? I suppose it's useful to link to all of them in one place, but then that has the problem of OSSU sort of endorsing them. Do we just link to anyone who decides to make such a thing regardless of how good it is? Are we making value judgements? Also, would this contribute to analysis paralysis? 

As for Alternative 1, I'm not sure what that would look like so I can't comment. Alternative 2 I am obviously against.

**Comment by waciumawanjohi** [comment]
The period for comments has closed. All comments received supported the proposed change.
